Add plugin: packUp4AI

shuxueshuxue opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin:

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [ ] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo commented:
["authorUrl": "",](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/manifest.json#L9), ["helpUrl": "",](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/manifest.json#L10)
If you don't have such a link remove this line.

[console.log(`Loading ${PLUGIN_NAME}`);](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L17), [console.log("Loading D3.js from CDN...");](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L295), [console.log("D3.js loaded successfully from CDN");](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L301), [console.log(`Starting collection with depth ${this.plugin.settings.maxDepth}`);](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L512), [console.log(`Collection completed: ${data.length} notes`);](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L534), [console.log("Collection changed, updating visualization");](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L538), [console.log("Collection unchanged, skipping visualization refresh");](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L541), [console.log(`Starting collection from ${startFile.path} with depth ${maxDepth}`);](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L58), [console.log(`Collected ${collected.length} notes`);](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L115)
Please avoid unnecessary logging. If you need it for debugging purposes, add a check to only log during development.

[id: 'open-packup4ai-notes',](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L24)
Don't include the plugin id in the command id, Obsidian will make sure that there are no conflicts with other plugins.

[name: 'Open packup4AI Notes',](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L25)
The name of this command doesn't make sense, your command doesn't open notes.

[const MODAL_TITLE = 'packup4AI Notes Explorer';](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L6), [name: 'Open packup4AI Notes',](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L25), [text: 'Collect & Copy to Clipboard',](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L433), [text: 'Collect & Save to File'](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L452), [pane.createEl('h3', { text: 'Collection Settings' });](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L610), [text: `Starting Note: ${this.currentFile.basename}`](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L614), [.setName('Collection Depth');](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L619), [.setName('Include Backlinks')](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L648), [.setName('Output Filename')](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L663), [pane.createEl('h3', { text: 'Note Relationship Graph' });](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L705), [.setName("Max Collection Depth")](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L328), [.setName("Include Backlinks")](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L340), [.setName("Default Output Filename")](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L350), [.setName("Excluded Files/Folders")](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L361)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[callback: () => {](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L26)
This should be a [conditional command](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands), with the condition being an active file.

[const cleanText = text.replace(/^---\s*[\s\S]*?---\s*/, '').trim();](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L165), [const clean = item.content](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L253-L255)
Use [`getFrontMatterInfo`](https://docs.obsidian.md/Reference/TypeScript+API/getFrontMatterInfo) to get information about the frontmatter of a file, including whether there is a frontmatter block, the offsets of where it starts and ends, and the frontmatter text.

[this.plugin.settings.outputFile = value.endsWith('.md') ? value : `${value}.md`;](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L668), [this.plugin.settings.outputFile = value.endsWith('.md') ? value : `${value}.md`;](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L356), [this.plugin.settings.excludePaths = value](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L367-L370)
[Use `normalizePath` to clean up user defined paths](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+%60normalizePath()%60+to+clean+up+user-defined+paths)

[containerEl.createEl("h2", { text: `${PLUGIN_NAME} Settings` });](https://github.com/shuxueshuxue/PackUp4AI/blob/ef80d56211e9bb927672396110ce2f17322d4e2a/main.js#L325)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.) And for the future, [Only use headings under settings if you have more than one section.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)and For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

---

Please consider including D3 in your plugin instead of downloading it every time your plugin is enabled, to enable your plugin to work offline.

To make the manual review process and future plugin maintenance more efficient, we recommend that you split your code into multiple files and move all source code files into a `src` directory. You can use a bundling tool such as `esbuild` to bundle your code into a single `main.js` file for releases. See [obsidian-sample-plugin](https://github.com/obsidianmd/obsidian-sample-plugin) for an example.

shuxueshuxue commented:
All Issues Addressed

  1. Empty URLs in manifest.json
  - Removed empty "authorUrl": "", and "helpUrl": "", fields from manifest.json

  2. No unnecessary console logging
  - Added process.env.NODE_ENV === 'development' checks around all console.log statements

  3. Command should be conditional
  - Changed from callback to checkCallback that only enables when active file exists

  4. Manual frontmatter removal
  - Replaced all regex-based frontmatter removal with getFrontMatterInfo

  5. Command ID includes plugin prefix
  - Fixed: 'open-packup4ai-notes' → 'collect-notes'

  6. Command name now makes sense
  -  'Open packup4AI Notes' → 'Collect notes for AI context'

  7. UI text not in sentence case
  - Updated all UI elements to sentence case:
    - "Collection Settings" → "Collection settings"
    - "Include Backlinks" → "Include backlinks"
    - "Output Filename" → "Output filename"
    - "Collect & Copy to Clipboard" → "Collect & copy to clipboard"
    - And all other instances

  8. Path normalization missing
  - Fixed: Added normalizePath calls for all user input paths

  9. Settings tab heading structure
  - Removed top-level `<h2>` heading from settings tab per guidelines

  10. D3 CDN dependency
  - local D3 inclusion

Zachatoo commented:
[const d3Content = await this.app.vault.adapter.read('.obsidian/plugins/packup4ai/d3.min.js');](https://github.com/shuxueshuxue/PackUp4AI/blob/3f8827dadc5cb09577838818010745fce72647cb/main.js#L317)
[Only the main.js, manifest.json, and styles.css (optional) files will be downloaded by Obsidian from your release](https://docs.obsidian.md/Plugins/Releasing/Submit+your+plugin#Step+2+Create+a+release). This is problem because your plugin relies on additional files in your release. You can use a bundling tool such as `esbuild` to bundle your code into a single `main.js` file for releases. See [obsidian-sample-plugin](https://github.com/obsidianmd/obsidian-sample-plugin) for an example.

When your plugin is unloaded or uninstalled, `window.d3` still exists. Your plugin should clean up after itself in the [`Plugin.onunload()`](https://docs.obsidian.md/Reference/TypeScript+API/Component/onunload) method.

---

Sorry about the wait, looks like the bot doesn't think you made one of the earlier requested changes but you did.

shuxueshuxue commented:
Fixed as requested.

Zachatoo commented:
[main.js](https://github.com/shuxueshuxue/PackUp4AI/blob/82d8445a14abdb1c4366eb91151ece89b75205b6/.gitignore#L3)
Why is this being ignored from git tracking? Please re-add your main.js to git tracking.

[console.log('D3 loaded successfully');](https://github.com/shuxueshuxue/PackUp4AI/blob/82d8445a14abdb1c4366eb91151ece89b75205b6/scripts/build.js#L40)
Please avoid unnecessary logging. If you need it for debugging purposes, add a check to only log during development.

[const bundledContent = // Obsidian Plugin with D3](https://github.com/shuxueshuxue/PackUp4AI/blob/82d8445a14abdb1c4366eb91151ece89b75205b6/scripts/build.js#L34-L49)
This technically works, but it is best practice to load resources in your plugin's `onload()` and clean them up in your plugin's `onunload()`. You're loading d3 before your plugin's `onload()` gets called. Please change this.
You should be able to use a bundler such as `esbuild` and simply import from your minified d3 module and the bundler will do the magic of pulling in that code appropriately for you.

shuxueshuxue commented:
Now using esbuild as the bundler.
D3 is dynamically imported in `onload()`. [onload](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L23)
Removed unnecessary console logging. [logging](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L29)

Thanks for the review.

Zachatoo commented:
Please remove the generated main.js file from your repo. You should still keep the src/main.js file, just remove the main.js file from the root of your repo.

[window.d3 = this.d3;](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L27), [if (window.d3) {](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L77-L79), [const d3 = window.d3;](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L769), [if (!window.d3) {](https://github.com/shuxueshuxue/PackUp4AI/blob/313f1231159484df51e22a4a46aa3a6f789239f4/src/main.js#L759)
You should not use global variables whenever possible, read here why: http://wiki.c2.com/?GlobalVariablesAreBad
You can instead pass `d3` or `this.d3` to your modal. Or, import `d3` in your modal's `unload()` method, since your plugin class doesn't use it.

shuxueshuxue commented:
Fixed.
- Removed global variable.
- Removed generated main.js file.
